From 564830719be2017c4953589d50f21a9e856a4ecc Mon Sep 17 00:00:00 2001
From: Chen Qi <Qi.Chen@windriver.com>
Date: Thu, 21 Feb 2019 16:23:24 +0800
Subject: [PATCH] binfmt: Don't install dependency links at install time for
 the binfmt services

use [Install] blocks so that they get created when the service is enabled
like a traditional service.

The [Install] blocks were rejected upstream as they don't have a way to
"enable" it on install without static symlinks which can't be disabled,
only masked. We however can do that in a postinst.

Upstream-Status: Denied

Signed-off-by: Ross Burton <ross.burton@intel.com>
Signed-off-by: Khem Raj <raj.khem@gmail.com>
Signed-off-by: Chen Qi <Qi.Chen@windriver.com>
[rebased for systemd 243]
Signed-off-by: Scott Murray <scott.murray@konsulko.com>

---
 units/meson.build                       | 6 ++----
 units/proc-sys-fs-binfmt_misc.automount | 3 +++
 units/systemd-binfmt.service.in         | 4 ++++
 3 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/units/meson.build b/units/meson.build
index ea91f0cc9ea7..25186f88dfeb 100644
--- a/units/meson.build
+++ b/units/meson.build
@@ -52,8 +52,7 @@ units = [
         ['poweroff.target',                     '',
          'runlevel0.target'],
         ['printer.target',                      ''],
-        ['proc-sys-fs-binfmt_misc.automount',   'ENABLE_BINFMT',
-         'sysinit.target.wants/'],
+        ['proc-sys-fs-binfmt_misc.automount',   'ENABLE_BINFMT'],
         ['proc-sys-fs-binfmt_misc.mount',       'ENABLE_BINFMT'],
         ['reboot.target',                       '',
          'runlevel6.target ctrl-alt-del.target'],
@@ -161,8 +160,7 @@ in_units = [
         ['rc-local.service',                     'HAVE_SYSV_COMPAT'],
         ['rescue.service',                       ''],
         ['systemd-backlight@.service',           'ENABLE_BACKLIGHT'],
-        ['systemd-binfmt.service',               'ENABLE_BINFMT',
-         'sysinit.target.wants/'],
+        ['systemd-binfmt.service',               'ENABLE_BINFMT'],
         ['systemd-bless-boot.service',           'ENABLE_EFI HAVE_BLKID'],
         ['systemd-boot-check-no-failures.service', ''],
         ['systemd-coredump@.service',            'ENABLE_COREDUMP'],
diff --git a/units/proc-sys-fs-binfmt_misc.automount b/units/proc-sys-fs-binfmt_misc.automount
index 30a6bc991844..4231f3b70fe9 100644
--- a/units/proc-sys-fs-binfmt_misc.automount
+++ b/units/proc-sys-fs-binfmt_misc.automount
@@ -18,3 +18,6 @@ ConditionPathIsReadWrite=/proc/sys/
 
 [Automount]
 Where=/proc/sys/fs/binfmt_misc
+
+[Install]
+WantedBy=sysinit.target
diff --git a/units/systemd-binfmt.service.in b/units/systemd-binfmt.service.in
index e54e95e11d5d..372a598614d3 100644
--- a/units/systemd-binfmt.service.in
+++ b/units/systemd-binfmt.service.in
@@ -14,6 +14,7 @@ Documentation=https://www.kernel.org/doc/html/latest/admin-guide/binfmt-misc.htm
 Documentation=https://www.freedesktop.org/wiki/Software/systemd/APIFileSystems
 DefaultDependencies=no
 Conflicts=shutdown.target
+Wants=proc-sys-fs-binfmt_misc.automount
 After=proc-sys-fs-binfmt_misc.automount
 After=proc-sys-fs-binfmt_misc.mount
 Before=sysinit.target shutdown.target
@@ -30,3 +31,6 @@ RemainAfterExit=yes
 ExecStart=@rootlibexecdir@/systemd-binfmt
 ExecStop=@rootlibexecdir@/systemd-binfmt --unregister
 TimeoutSec=90s
+
+[Install]
+WantedBy=sysinit.target
